src/Microsoft.Health.Fhir.Core/Features/Security/Authorization/RoleBasedFhirAuthorizationService.cs

```diff
@@ -53,7 +54,14 @@ public ValueTask<DataActions> CheckAccess(DataActions dataActions, CancellationT
                 }
             }
 
-            return new ValueTask<DataActions>(dataActions & permittedDataActions);
+            if (applySMARTAcess)
```

Comment:
This is little confusing with the class name or may be its just me.
RoleBasedFhirAuthorizationService.cs should just return role based auth but this is also internally calling Smart based auth. Could we refactor IAuthoriziationService and add two methods CheckRoleBasedAccess and CheckSmartScopeBasedAccess and add corresponding impl?
And probably we can rename class to general one probably FhirAuthorizationService?

Reply:
I agree that we are overloading this method.  In the context of the SmatUser role, applying clinical scopes is part of the access rules for that role, but the idea that access control is becoming more than simply checking the RBAC actions is accurate.  We debated the amount of change we wanted to make right now, and as the CheckAccess method is being invoked in all the places we also want to check for clinical scopes, it was a good fit, for now.  As this may become more complex, we will want to follow something like your suggestion.